### PR TITLE
Basic OTel support

### PR DESCRIPTION
**Pull request by antonpirker**
Adding basic OpenTelementry (OTel) support to the Sentry SDK:
* Adding a OTel `SpanProcessor` that can receive spans form OTel and then convert them into Sentry Spans and send them to Sentry.
* Adding a OTel `Propagator` that can receive and propagate trace headers (Baggage) to keep distributed tracing intact.

### CODE REVIEW

**Comment by sl0thentr0py** [comment]
we need a line in setup.py in `extras_require`

File: sentry_sdk/integrations/opentelemetry/span_processor.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+OPEN_TELEMETRY_CONTEXT = "otel"
+
+
+class SentrySpanProcessor(SpanProcessor):  # type: ignore
+    """
+    Converts OTel spans into Sentry spans so they can be sent to the Sentry backend.
+    """
+
+    # The mapping from otel span ids to sentry spans
+    otel_span_map = {}  # type: Dict[str, Union[Transaction, OTelSpan]]
```
Comment (sl0thentr0py) [inline sentry_sdk/integrations/opentelemetry/span_processor.py:40]:
```suggestion
    otel_span_map = {}  # type: Dict[str, Union[Transaction, SentrySpan]]
```

File: sentry_sdk/integrations/opentelemetry/span_processor.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+    def __new__(cls):
+        # type: () -> SentrySpanProcessor
+        if not hasattr(cls, "instance"):
+            cls.instance = super(SentrySpanProcessor, cls).__new__(cls)
+
+        return cls.instance
+
+    def on_start(self, otel_span, parent_context=None):
+        # type: (OTelSpan, SpanContext) -> None
```
Comment (sl0thentr0py) [inline sentry_sdk/integrations/opentelemetry/span_processor.py:50]: this is context, not span context
```suggestion
        # type: (OTelSpan, Optional[Context]) -> None
```

File: sentry_sdk/integrations/opentelemetry/span_processor.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        # type: (OTelSpan) -> None
+        hub = Hub.current
+        if not hub:
+            return
+
+        if hub.client and hub.client.options["instrumenter"] != INSTRUMENTER.OTEL:
+            return
+
+        # Break infinite http requests to Sentry are caught by OTel and send again to Sentry.
+        otel_span_url = otel_span.attributes.get(SpanAttributes.HTTP_URL, None)
```
Comment (sl0thentr0py) [inline sentry_sdk/integrations/opentelemetry/span_processor.py:96]: is it also not possible to do this in on_start like in java? 

can you link the to the http instrumentation in otel?

File: sentry_sdk/integrations/opentelemetry/span_processor.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+        return trace_data
+
+    def _update_span_with_otel_data(self, sentry_span, otel_span):
+        # type: (SentrySpan, OTelSpan) -> None
+        """
+        Convert OTel span data and update the Sentry span with it.
+        This should eventually happen on the server when ingesting the spans.
+        """
+        for key in otel_span.attributes:
```
Comment (sl0thentr0py) [inline sentry_sdk/integrations/opentelemetry/span_processor.py:171]: why not?
```suggestion
        for key, val in otel_span.attributes:
```

File: sentry_sdk/integrations/opentelemetry/span_processor.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+            val = otel_span.attributes[key]
+            sentry_span.set_data(key, val)
+
```
Comment (sl0thentr0py) [inline sentry_sdk/integrations/opentelemetry/span_processor.py:174]: missing 
```python
sentry_span.set_data("otel.kind", otel_span.kind)
```

**Review by sl0thentr0py — COMMENTED** [review]
reviewed `SpanProcessor`

File: sentry_sdk/integrations/opentelemetry/span_processor.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    def on_end(self, otel_span):
+        # type: (OTelSpan) -> None
+        hub = Hub.current
+        if not hub:
+            return
+
+        if hub.client and hub.client.options["instrumenter"] != INSTRUMENTER.OTEL:
+            return
+
```
Comment (sl0thentr0py) [inline sentry_sdk/integrations/opentelemetry/span_processor.py:94]: need another check here
```python
if not otel_span.context.is_valid:
    return
```

File: sentry_sdk/integrations/opentelemetry/propagator.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+    def inject(self, carrier, context=None, setter=default_setter):
+        # type: (CarrierT, Optional[Context], Setter) -> None
+        if context is None:
+            context = get_current()
+
+        current_span = trace.get_current_span(context)
+        span_id = trace.format_span_id(current_span.context.span_id)
+
+        from sentry_sdk.integrations.opentelemetry.span_processor import (
```
Comment (sl0thentr0py) [inline sentry_sdk/integrations/opentelemetry/propagator.py:99]: why import inside the function and not at the top?

File: sentry_sdk/integrations/opentelemetry/propagator.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        span = NonRecordingSpan(span_context)
+        modified_context = trace.set_span_in_context(span, context)
+        return modified_context
+
+    def inject(self, carrier, context=None, setter=default_setter):
+        # type: (CarrierT, Optional[Context], Setter) -> None
+        if context is None:
+            context = get_current()
+
+        current_span = trace.get_current_span(context)
```
Comment (sl0thentr0py) [inline sentry_sdk/integrations/opentelemetry/propagator.py:96]: need a 
need another check here
```python
if not current_span.context.is_valid:
    return
```

File: sentry_sdk/integrations/opentelemetry/propagator.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        )
+
+        span_map = SentrySpanProcessor().otel_span_map
+        sentry_span = span_map.get(span_id, None)
+        if not sentry_span:
+            return
+
+        setter.set(carrier, SENTRY_TRACE_HEADER_NAME, sentry_span.to_traceparent())
+
+        baggage = hasattr(sentry_span, "get_baggage") and sentry_span.get_baggage()
```
Comment (sl0thentr0py) [inline sentry_sdk/integrations/opentelemetry/propagator.py:110]:
```suggestion
        baggage = sentry_span.containing_transaction.get_baggage().serialize()
```

**Review by sl0thentr0py — CHANGES_REQUESTED** [review]
alright did propagator too, will do another round once all this is changed

File: sentry_sdk/tracing.py [hunk 10]
```diff
@@ -869,7 +869,7 @@ def __exit__(self, ty, value, tb):
 
     def start_child(self, instrumenter=INSTRUMENTER.SENTRY, **kwargs):
         # type: (str, **Any) -> Any
```
[the remark this answers is not in the document]
Comment (antonpirker) [reply sentry_sdk/tracing.py]: Done

File: sentry_sdk/integrations/opentelemetry/span_processor.py (hunk 3 above)
Comment (antonpirker) [reply sentry_sdk/integrations/opentelemetry/span_processor.py]: Yes it is possible. I changed that

File: sentry_sdk/integrations/opentelemetry/propagator.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+        sentry_trace = getter.get(carrier, SENTRY_TRACE_HEADER_NAME)
+        if not sentry_trace:
+            return context
+
+        sentrytrace = extract_sentrytrace_data(sentry_trace[0])
+        if not sentrytrace:
+            return context
+
+        sentry_trace_data = (
```
[the remark this answers is not in the document]
Comment (antonpirker) [reply sentry_sdk/integrations/opentelemetry/propagator.py]: You are right. I wanted to make it as similar to Ruby as possible, but I guess this does not make sense here. :-) Changed it.

File: sentry_sdk/integrations/opentelemetry/span_processor.py (hunk 3 above)
Comment (sl0thentr0py) [reply sentry_sdk/integrations/opentelemetry/span_processor.py]: for documentation, this is the place where this is set in otel
https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py#L195-L200

File: sentry_sdk/integrations/opentelemetry/propagator.py (hunk 7 above)
Comment (antonpirker) [reply sentry_sdk/integrations/opentelemetry/propagator.py]: Circular imports. Fixed it.

File: sentry_sdk/integrations/opentelemetry/propagator.py (hunk 9 above)
Comment (antonpirker) [reply sentry_sdk/integrations/opentelemetry/propagator.py]: Done.

Comment (antonpirker) [reply sentry_sdk/integrations/opentelemetry/propagator.py]: (but slightly different)

**Review by vladanpaunovic — APPROVED** [review]

**Pull request merged**